### PR TITLE
CLI arguments

### PR DESCRIPTION
**Pull request by mikiw**
## Usage related changes

- Added CLI integration for: --host, --port, --gas-price, --timeout, --chain-id

## Development related changes

- Const `CHAIN_ID` was removed, for test we ar using `StarknetChainId::TestNet`

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet-server/src/cli.rs [hunk 1]
```diff
@@ -43,6 +79,16 @@ impl Args {
...
                 .expect("Invalid value for initial balance"), // TODO doesn't exit nicely
+            host: self.host.clone(),
+            port: self.port,
+            timeout: self.timeout,
+            gas_price: self.gas_price,
+            chain_id: match self.chain_id.as_str() {
+                "MAINNET" => StarknetChainId::MainNet,
+                "TESTNET" => StarknetChainId::TestNet,
+                "TESTNET2" => StarknetChainId::TestNet2,
+                _ => StarknetChainId::TestNet,
```
Comment (FabijanC) [inline crates/starknet-server/src/cli.rs:90]: This is wrong, it means if I do `--chain-id lmao-i-am-an-invalid-chain-id`, it will default to TESTNET.

**Review by FabijanC — CHANGES_REQUESTED** [review]
Notice a difference between the failure of:
```shell
$ cargo run -- --host lmao
```
and
```shell
$ cargo run -- --port lmao
```

Ideally all of our startup failures would look as cool as the latter. This is mentioned in some comments.

File: crates/starknet-server/src/cli.rs (hunk 1 above)
Comment (mikiw) [reply crates/starknet-server/src/cli.rs:90]: I thought that it's expected behavior but I'll raise exception in that case

File: crates/starknet-server/src/cli.rs [hunk 2]
```diff
@@ -1,33 +1,69 @@
...
+    #[arg(long = "timeout")]
+    #[arg(value_name = "TIMEOUT")]
+    #[arg(default_value = "120")]
+    #[arg(help = "Specify the server timeout in seconds;")]
+    timeout: u16,
+
+    // Gas price in wei.
+    #[arg(long = "gas-price")]
+    #[arg(value_name = "GAS_PRICE")]
+    #[arg(default_value = "100000000")]
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/src/cli.rs:57]: it was taken from devnet in python

File: crates/starknet-server/src/cli.rs [hunk 3]
```diff
@@ -1,33 +1,69 @@
...
+    // Host address.
+    #[arg(long = "host")]
+    #[arg(value_name = "DEVNET_HOST")]
+    #[arg(default_value = "127.0.0.1")]
+    #[arg(help = "Specify the address to listen at;")]
+    host: String,
+
+    // Port number.
+    #[arg(long = "port")]
+    #[arg(value_name = "DEVNET_PORT")]
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/src/cli.rs]: Fixed in 588b138cdd08d488169824ad19a9d684712aca25

File: crates/starknet-server/src/cli.rs [hunk 4]
```diff
@@ -1,33 +1,69 @@
...
+    #[arg(long = "host")]
+    #[arg(value_name = "HOST")]
+    #[arg(default_value = "127.0.0.1")]
+    #[arg(help = "Specify the address to listen at;")]
+    host: String,
+
+    // Port number.
+    #[arg(long = "port")]
+    #[arg(value_name = "PORT")]
+    #[arg(default_value = "5050")]
```
Comment (taco-paco) [inline crates/starknet-server/src/cli.rs:43]: `default_value_t` shall be used for data types.
https://github.com/clap-rs/clap/discussions/3640

File: crates/starknet-server/src/main.rs [hunk 5]
```diff
@@ -53,15 +53,8 @@ async fn main() -> Result<(), anyhow::Error> {
...
-
-    // configure server
-    let port = env::var("DEVNET_PORT")
-        .expect("DEVNET_PORT must be set")
-        .parse::<u16>()
-        .expect("DEVNET_PORT must be a valid port number");
-
-    let host = IpAddr::V4(Ipv4Addr::LOCALHOST);
-    let mut addr = SocketAddr::new(host, port);
+    let host = IpAddr::from_str(starknet_config.host.as_str()).unwrap();
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet-server/src/main.rs]: Resolving this conversation, but mentioned in https://github.com/0xSpaceShard/starknet-devnet-rs/issues/34

File: crates/starknet/src/lib.rs [hunk 6]
```diff
@@ -238,6 +250,11 @@ mod tests {
             seed: 123,
             total_accounts: 3,
             predeployed_accounts_initial_balance: 100.into(),
+            host: String::from("127.0.0.1"),
+            port: 5050,
```
Comment (taco-paco) [inline crates/starknet/src/lib.rs:254]: Can be extracted in default consts and reused in cli.rs

File: crates/starknet-server/src/main.rs [hunk 7]
```diff
@@ -53,15 +54,8 @@ async fn main() -> Result<(), anyhow::Error> {
...
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/src/main.rs]: it was fixed in 7ca6b0fa2fcf6f131de1d59da374689a738dd71d

File: crates/server/Cargo.toml [hunk 8]
```diff
@@ -26,4 +26,7 @@ async-trait = { workspace = true }
 # misc
 serde_json = { workspace = true }
 serde = { workspace = true }
-thiserror = { workspace = true }
\ No newline at end of file
+thiserror = { workspace = true }
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/server/Cargo.toml:29]: `thiserror` was not introduced by myself, we can create improvement task with that if you this it's necessary

File: crates/starknet/src/lib.rs (hunk 6 above)
Comment (mikiw) [reply crates/starknet/src/lib.rs]: I'm not sure how to do it :( a little help?

File: crates/starknet-server/src/cli.rs [hunk 9]
```diff
@@ -1,33 +1,71 @@
...
+use starknet_in_rust::definitions::block_context::StarknetChainId;
 use starknet_types::num_bigint::BigUint;
 
 /// Run a local instance of Starknet Devnet
 #[derive(Parser, Debug)]
 #[command(author, version, about, long_about = None)]
 #[command(propagate_version = true)]
 pub(crate) struct Args {
-    /// Number of accounts
+    /// Number of accounts.
```
Comment (taco-paco) [inline crates/starknet-server/src/cli.rs:13]: I think we don't use `.` in other places. at the end

Comment (mikiw) [reply crates/starknet-server/src/cli.rs]: Fixed in 796f389d0c66a19de306a91b29b072fccc9fb4f4

File: crates/starknet/src/lib.rs (hunk 6 above)
Comment (mikiw) [reply crates/starknet/src/lib.rs]: solved in recent commits

File: crates/server/src/builder.rs [hunk 10]
```diff
@@ -96,7 +99,9 @@ impl<TJsonRpcHandler: RpcHandler, THttpApiHandler: Clone + Send + Sync + 'static
             svc = svc.layer(Extension(self.http_api_handler.unwrap()));
         }
 
-        svc = svc.layer(TraceLayer::new_for_http());
+        svc = svc
+            .layer(TraceLayer::new_for_http())
+            .layer(TimeoutLayer::new(Duration::from_secs(starknet_config.timeout.into())));
```
Comment (taco-paco) [inline crates/server/src/builder.rs:104]: Why do we need this layer?

Comment (mikiw) [reply crates/server/src/builder.rs:104]: to set a timeout, I discussed that `TimeoutLayer` with @marioiordanov

**Review by taco-paco — APPROVED** [review]

File: crates/server/src/builder.rs (hunk 10 above)
Comment (taco-paco) [reply crates/server/src/builder.rs:104]: What this timeout used for?

Comment (mikiw) [reply crates/server/src/builder.rs:104]: in devnet in python, sometimes execution was longer than 30, or 60 seconds so there was a need to exceed timeout and make it a CLI parameter

**Review by FabijanC — APPROVED** [review]

**Pull request merged**